File: app/src/main/java/com/saltwater/baseproject/utils/ScreenAdaptationUtil.java
package com.saltwater.baseproject.utils;

import android.app.Activity;
import android.app.Application;
import android.content.ComponentCallbacks;
import android.content.res.Configuration;
import android.support.annotation.NonNull;
import android.util.DisplayMetrics;

/**
 * <pre>
 *     author : wenxin
 *     e-mail : wenxin2233@outlook.com
 *     time   : 2018/07/11
 *     desc   : 简单的屏幕适配方案
 *     version: 1.0
 * </pre>
 */
public class ScreenAdaptationUtil {
    private static float sNoncompatDensity;
    private static float sNoncompatScaledDensity;
    private static int sNoncompatDensityDpi;

    private static final int sWidth = 375;//375 设计图的宽度dp 根据宽度适配
    private static final int sHeight = 667;//667 设计图的高度dp 根据高度适配


    /**
     * 默认按照宽度适配
     *
     * @param activity
     * @param application
     * @param isWidth     是否按照宽度适配，false按照高度适配
     */
    public static void setCustomDensity(@NonNull Activity activity, @NonNull final Application application, boolean isWidth) {
        final DisplayMetrics appdisplayMetrics = application.getResources().getDisplayMetrics();
        if (sNoncompatDensity == 0) {
            sNoncompatDensity = appdisplayMetrics.density;
            sNoncompatScaledDensity = appdisplayMetrics.scaledDensity;
            sNoncompatDensityDpi = appdisplayMetrics.densityDpi;
            application.registerComponentCallbacks(new ComponentCallbacks() {
                @Override
                public void onConfigurationChanged(Configuration configuration) {
                    if (configuration != null && configuration.fontScale > 0) {
                        sNoncompatScaledDensity = application.getResources().getDisplayMetrics().scaledDensity;
                    }
                }

                @Override
                public void onLowMemory() {

                }
            });
        }
        final float targetDesity;
        if (isWidth) {
            targetDesity = (float) appdisplayMetrics.widthPixels / sWidth;
        } else {
            targetDesity = (float) appdisplayMetrics.heightPixels / sHeight;
        }
        final float targetScaleDesity = targetDesity * (sNoncompatScaledDensity / sNoncompatDensity);
        final int targetDesityDpi = (int) (160 * targetDesity);

        appdisplayMetrics.density = targetDesity;
        appdisplayMetrics.scaledDensity = targetScaleDesity;
        appdisplayMetrics.densityDpi = targetDesityDpi;

        final DisplayMetrics activityDisplayMetrics = activity.getResources().getDisplayMetrics();
        activityDisplayMetrics.density = targetDesity;
        activityDisplayMetrics.scaledDensity = targetScaleDesity;
        activityDisplayMetrics.densityDpi = targetDesityDpi;
    }

    public static void resetDensity(@NonNull Activity activity, @NonNull final Application application) {
        DisplayMetrics appdisplayMetrics = application.getResources().getDisplayMetrics();
        appdisplayMetrics.density = sNoncompatDensity;
        appdisplayMetrics.scaledDensity = sNoncompatScaledDensity;
        appdisplayMetrics.densityDpi = sNoncompatDensityDpi;

        final DisplayMetrics activityDisplayMetrics = activity.getResources().getDisplayMetrics();
        activityDisplayMetrics.density = sNoncompatDensity;
        activityDisplayMetrics.scaledDensity = sNoncompatScaledDensity;
        activityDisplayMetrics.densityDpi = sNoncompatDensityDpi;
    }
}

File: ModuleCommon/src/main/java/com/saltwater/modulecommon/utils/ScreenAdaptationUtil.kt
package com.saltwater.modulecommon.utils

import android.app.Activity
import android.app.Application
import android.content.ComponentCallbacks
import android.content.res.Configuration

/**
 * <pre>
 * author : wenxin
 * e-mail : wenxin2233@outlook.com
 * time   : 2018/07/11
 * desc   : 简单的屏幕适配方案,通过把density设置为统一进行适配
 * version: 1.0
</pre> *
 */
object ScreenAdaptationUtil {
    private var density = 0f
    private var scaledDensity = 0f
    private var densityDpi = 0

    private const val width = 375//375 设计图的宽度dp 根据宽度适配
    private const val height = 667//667 设计图的高度dp 根据高度适配


    /**
     * 默认按照宽度适配
     *
     * @param activity
     * @param application
     * @param isWidth     是否按照宽度适配，false按照高度适配
     */
    fun setCustomDensity(activity: Activity, application: Application, isWidth: Boolean) {
        val airplayMetrics = application.resources.displayMetrics
        if (density == 0f) {
            density = airplayMetrics.density
            scaledDensity = airplayMetrics.scaledDensity
            densityDpi = airplayMetrics.densityDpi
            application.registerComponentCallbacks(object : ComponentCallbacks {
                override fun onConfigurationChanged(configuration: Configuration?) {
                    if (configuration != null && configuration.fontScale > 0) {
                        scaledDensity = application.resources.displayMetrics.scaledDensity
                    }
                }

                override fun onLowMemory() {

                }
            })
        }
        val targetDeity: Float
        if (isWidth) {
            targetDeity = airplayMetrics.widthPixels.toFloat() / width
        } else {
            targetDeity = airplayMetrics.heightPixels.toFloat() / height
        }
        val targetScaleDeity = targetDeity * (scaledDensity / density)
        val targetDensityDpi = (160 * targetDeity).toInt()

        airplayMetrics.density = targetDeity
        airplayMetrics.scaledDensity = targetScaleDeity
        airplayMetrics.densityDpi = targetDensityDpi

        val activityDisplayMetrics = activity.resources.displayMetrics
        activityDisplayMetrics.density = targetDeity
        activityDisplayMetrics.scaledDensity = targetScaleDeity
        activityDisplayMetrics.densityDpi = targetDensityDpi
    }

    //重设DisplayMetrics
    fun resetDensity(activity: Activity, application: Application) {
        val displayMetrics = application.resources.displayMetrics
        displayMetrics.density = density
        displayMetrics.scaledDensity = scaledDensity
        displayMetrics.densityDpi = densityDpi

        val activityDisplayMetrics = activity.resources.displayMetrics
        activityDisplayMetrics.density = density
        activityDisplayMetrics.scaledDensity = scaledDensity
        activityDisplayMetrics.densityDpi = densityDpi
    }
}
